### PR TITLE
Add ProducerSideServiceImpl and ProducerSideServer

### PR DESCRIPTION
**Pull request by dpallotti**
For the communication between producers and `OrbitService`. Bug: http://b/174028631.

Synchronization mess part 1: I'm aware that the synchronization in `ProducerSideServiceImpl` is a bit tricky and especially not so easy on the reader. I'm not really sure how to improve it, but I tried to test it quite a bit.

### CODE REVIEW

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received, std::atomic<bool>* exit_send_commands_thread) {
+  // As a result of initializing prev_capture_status to kCaptureFinished,
+  // an initial StartCaptureCommand is sent
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStarted,
+  // and an initial StopCaptureCommand is sent (with little effect)
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  while (true) {
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:131]: A general description of that loop (or just the whole method), would be beneficial.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    // Wait for service_state_.capture_status to change or for service_state->exit_requested
+    // (the next iteration will handle the change).
+    // Use a timeout to periodically check (in the next iteration)
+    // for *terminate_send_commands_thread, set by ReceiveCommandsAndSendEvents.
+    static constexpr absl::Duration kCheckExitSendCommandsThreadInterval = absl::Seconds(1);
+    switch (service_state_.capture_status) {
+      case CaptureStatus::kCaptureStarted: {
+        service_state_mutex_.AwaitWithTimeout(absl::Condition(
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:195]: It seems that you need this super duplicated code (for all the three cases), as you need a function pointer in the AwaitWithTimout, which can't capture the state at the beginning. If so, could you add comment for that?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      absl::MutexLock lock{&service_state_mutex_};
+      if (service_state_.exit_requested) {
+        break;
+      }
+    }
+
+    switch (request.event_case()) {
+      case orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest::kBufferedCaptureEvents: {
+        absl::MutexLock lock{&capture_event_buffer_mutex_};
+        if (capture_event_buffer_ != nullptr) {
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:250]: In what cases would the buffer be nullptr? Is that reasonable? Should this be rather a check, or do we want to log some error in this case?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        }
+      } break;
+
+      case orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest::kAllEventsSent: {
+        LOG("Received AllEventsSent from CaptureEventProducer");
+        absl::MutexLock lock{&service_state_mutex_};
+        switch (service_state_.capture_status) {
+          case CaptureStatus::kCaptureStarted: {
+            ERROR("CaptureEventProducer sent AllEventsSent while still capturing");
+            if (!*all_events_sent_received) {
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:264]: That would benefit from a brief comment.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+          }
+        }
+      } break;
+
+      case orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest::kAllEventsSent: {
+        LOG("Received AllEventsSent from CaptureEventProducer");
+        absl::MutexLock lock{&service_state_mutex_};
+        switch (service_state_.capture_status) {
+          case CaptureStatus::kCaptureStarted: {
+            ERROR("CaptureEventProducer sent AllEventsSent while still capturing");
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:263]: Can this actually happen?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+  }
+}
+
+void ProducerSideServiceImpl::ReceiveEventsThread(
+    grpc::ServerContext* /*context*/,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received) {
+  orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest request;
+  while (stream->Read(&request)) {
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImpl.cpp:239]: Shouldn't there also an ` || *all_events_sent_received`?

File: OrbitService/ProducerSideServiceImplTest.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      stream_;
+  std::thread read_thread_;
+};
+
+class MockCaptureEventBuffer : public CaptureEventBuffer {
+ public:
+  MOCK_METHOD(void, AddEvent, (orbit_grpc_protos::CaptureEvent && event), (override));
+};
+
+class ProducerSideServiceImplTest : public ::testing::Test {
```
Comment (florian-kuebler) [inline OrbitService/ProducerSideServiceImplTest.cpp:101]: Could you also add some tests for the corner cases (errors, no capture data, no AllEventsSend, etc.?).

**Review by florian-kuebler — CHANGES_REQUESTED** [review]
Thanks for that, overall looking very good. I was afraid of the synchronization, but it looks good from my side. But beside the minor commits, I really would like to see a couple of tests for the corner cases.

File: OrbitService/ProducerSideServiceImplTest.cpp (hunk 7 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImplTest.cpp:101]: Added some more tests, please have a look.

File: OrbitService/ProducerSideServiceImpl.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+#define ORBIT_SERVICE_PRODUCER_SIDE_SERVICE_IMPL_H_
+
+#include "CaptureStartStopListener.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "producer_side_services.grpc.pb.h"
+
+namespace orbit_service {
+
+class ProducerSideServiceImpl final : public orbit_grpc_protos::ProducerSideService::Service,
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitService/ProducerSideServiceImpl.h:15]: But there are additional methods. So I was thinking of a very very brief insight of the implementation and the additional methods. 
Regarding thread-safety, basically the same. Some rough ideas whats going on with the mutexes here would be nice.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                                ::orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream) {
+  LOG("A CaptureEventProducer has connected calling ReceiveCommandsAndSendEvents");
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.emplace(context);
+  }
+
+  // This will also be protected by service_state_mutex_.
+  bool all_events_sent_received = true;
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitService/ProducerSideServiceImpl.cpp:93]: Looks good now, thanks.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 3 above)
Comment (florian-kuebler) [reply OrbitService/ProducerSideServiceImpl.cpp:269]: Okay, thanks for explanation. I leave it up to you on whether to log errors here or not.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 4 above)
Comment (florian-kuebler) [reply OrbitService/ProducerSideServiceImpl.cpp:285]: Sounds good. Thanks!

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 5 above)
Comment (florian-kuebler) [reply OrbitService/ProducerSideServiceImpl.cpp:282]: I see. Thanks!

**Review by florian-kuebler — APPROVED** [review]
Perfect, thanks for the additional tests. LGTM now!

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+
+    // Wait (for a limited amount of time) for all producers to send AllEventsSent or to disconnect.
+    service_state_mutex_.AwaitWithTimeout(absl::Condition(
+                                              +[](ServiceState* service_state) {
+                                                return service_state->producers_remaining == 0 ||
+                                                       service_state->exit_requested;
+                                              },
+                                              &service_state_),
+                                          max_wait_for_all_events_sent_);
+    CHECK(service_state_.producers_remaining >= 0);
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:40]: == 0 || exit requested?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    for (grpc::ServerContext* context : server_contexts_) {
+      // This should cause blocking Reads on ServerReaderWriter to fail immediately.
+      context->TryCancel();
+    }
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:76]: maybe extract this to ClearEventBuffer?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+grpc::Status ProducerSideServiceImpl::ReceiveCommandsAndSendEvents(
+    ::grpc::ServerContext* context,
+    ::grpc::ServerReaderWriter< ::orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                                ::orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream) {
+  LOG("A CaptureEventProducer has connected calling ReceiveCommandsAndSendEvents");
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.emplace(context);
+  }
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:88]: SetContext?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+          command.mutable_start_capture_command();
+          if (!stream->Write(command)) {
+            ERROR("Sending StartCaptureCommand to CaptureEventProducer");
+            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
+            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
+            context->TryCancel();
+            return;
+          }
+          LOG("Sent StartCaptureCommand to CaptureEventProducer");
+        } break;
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:173]: nit: move break inside {}?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  // This loop keeps track of changes to service_state_.capture_status using
+  // conditional critical sections on service_state_mutex_ and updating prev_capture_status,
+  // and sends StartCaptureCommands and StopCaptureCommands accordingly.
+  // It exits when *exit_send_commands_thread is true, when service_state_.exit_requested is true,
+  // or when Write fails (because the producer disconnected or because the context was cancelled).
+  while (true) {
+    // This is set when ReceiveEventsThread has exited. At that point this thread should also exit.
+    if (*exit_send_commands_thread) {
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:145]: revieve_event_thread_exited?

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
+      switch (service_state_.capture_status) {
+        case CaptureStatus::kCaptureStarted: {
+          prev_capture_status = service_state_.capture_status;
+          ++service_state_.producers_remaining;
+          *all_events_sent_received = false;
+          service_state_mutex_.Unlock();
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:161]: Yes this is really hard to track. One thing you could do is to set prev_capture_status in L153, You can then use it in switch without locking. As for other fields... What would happen if you reistablished the lock here? Is there a possibility of race condition on them in other words do they have to be scoped with status read?

**Review by dimitry- — COMMENTED** [review]
Hi, sorry for delay in review - some preliminary comments. Still looking at the change.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 10 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:40]: That would be false in case of timeout. I could do something like `!condition_was_satisfied || service_state_.producers_remaining == 0 || service_state->exit_requested` (I think? Or something like that), but that actually goes beyond the point of this `CHECK`, as it's basically restating the condition, while I just wanted to have a `CHECK` on `service_state_.producers_remaining`.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 11 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:76]: I know this is a minor code duplication, but in this code I really want to avoid extracting code that locks mutexes outside of the already numerous (but necessary) methods that take those mutexes, as doing that distracts from the actual code flow as it requires jumping somewhere else. In other words in this complex synchronized code I want to prioritize following the flow (which is already hard enough) to normal code practices.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 12 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:88]: As above, I don't want to add more methods that take mutexes, even more here are there's no code duplication.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 13 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:173]: The break outside is something we do elsewhere and that I like. The idea is to separate the actual code from the `break` required by the old `switch` construct, which has nothing to do with the content of the braces.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 14 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp]: I guess it's better, done.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 10 above)
Comment (dimitry-) [reply OrbitService/ProducerSideServiceImpl.cpp:40]: maybe just remove the check in that case, I do not think it adds anything.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 15 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:161]: Hmmm, I don't think the mutex should be acquired twice, as `service_state_.capture_status` and `service_state_.producers_remaining` and also `all_events_sent_received` go together. The value assigned to the last two depends quite strongly on the current value of `service_state_.capture_status`. That shouldn't be relaxed only to make it easier to track `Unlock`s.

File: OrbitService/ProducerSideServer.cpp [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+  builder.RegisterService(&producer_side_service_);
+
+  server_ = builder.BuildAndStart();
+  if (server_ == nullptr) {
+    return false;
+  }
+
+  // When OrbitService runs as root, also allow non-root producers
+  // (e.g., the game) to communicate over the Unix domain socket.
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply OrbitService/ProducerSideServer.cpp:28]: yes - it is about todo.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 10 above)
Comment (dimitry-) [reply OrbitService/ProducerSideServiceImpl.cpp:40]: well, I was under impression that producers_remaining cannot go negative. but ok. you can keep the check if you want to make sure it is not negative.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 15 above)
Comment (dimitry-) [reply OrbitService/ProducerSideServiceImpl.cpp:161]: ok, then split the case into 2. One scoped on the mutex - and the second one doing communication based on prev_capture_status, it is going to be more readable and will have a clear mutex scope.

File: OrbitService/ProducerSideServer.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitService/ProducerSideServer.cpp:28]: Ok, I'm certainly happy to remove the TODO.

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 15 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:161]: You mean I should have two separate `switch`es?

Comment (dimitry-) [reply OrbitService/ProducerSideServiceImpl.cpp:161]: yes, switches, it makes code more readable, and mutex scope easier to follow.

Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp]: Makes sense, done.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 17]
```diff
@@ -241,9 +246,9 @@ void ProducerSideServiceImpl::SendCommandsThread(
                                                   },
                                                   &service_state_),
                                               kCheckExitSendCommandsThreadInterval);
-        service_state_mutex_.Unlock();
       } break;
     }
+    service_state_mutex_.Unlock();
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:251]: I wonder if this can be scoped somehow as well. Maybe check == case first with early continue?

Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp:251]: I don't understand the suggestion.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+      return;
+    }
+
+    service_state_mutex_.Lock();
+    if (service_state_.exit_requested) {
+      service_state_mutex_.Unlock();
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:155]: if you reverse this if - you can scope service_state_mutex_ better. I think.

Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp]: Done, it doesn't make the biggest difference but I guess it's a bit better.

Let me know if there's something else that needs to be addressed.

File: OrbitService/ProducerSideServiceImpl.cpp [hunk 19]
```diff
@@ -152,103 +152,104 @@ void ProducerSideServiceImpl::SendCommandsThread(
...
+      } break;
     }
+    prev_capture_status = service_state_.capture_status;
 
-    // Wait for service_state_.capture_status to change or for service_state->exit_requested
-    // (the next iteration will handle the change).
-    // Use a timeout to periodically check (in the next iteration)
-    // for *terminate_send_commands_thread, set by ReceiveCommandsAndSendEvents.
-    static constexpr absl::Duration kCheckExitSendCommandsThreadInterval = absl::Seconds(1);
+    service_state_mutex_.Unlock();
```
Comment (dimitry-) [inline OrbitService/ProducerSideServiceImpl.cpp:221]: And at this point you can use scoped lock I think: MutexLock lock (....)

**Review by dimitry- — APPROVED** [review]
otherwise lgtm

File: OrbitService/ProducerSideServiceImpl.cpp (hunk 19 above)
Comment (dpallotti) [reply OrbitService/ProducerSideServiceImpl.cpp]: Done.
